src/detect-tls-random.c

```diff
@@ -0,0 +1,296 @@
+    sigmatch_table[DETECT_AL_TLS_RANDOM].Setup = DetectTlsRandomSetup;
+    sigmatch_table[DETECT_AL_TLS_RANDOM].flags |= SIGMATCH_NOOPT | SIGMATCH_INFO_STICKY_BUFFER;
+
+    /* Register engine for Server random */
+    DetectAppLayerInspectEngineRegister2("tls.random", ALPROTO_TLS, SIG_FLAG_TOSERVER, 0,
+            DetectEngineInspectBufferGeneric, GetRandomData);
+
+    DetectAppLayerMpmRegister2("tls.random", SIG_FLAG_TOSERVER, 2, PrefilterGenericMpmRegister,
+            GetRandomData, ALPROTO_TLS, 0);
+
+    /* Register engine for Client random */
+    DetectAppLayerInspectEngineRegister2("tls.random", ALPROTO_TLS, SIG_FLAG_TOCLIENT, 0,
+            DetectEngineInspectBufferGeneric, GetRandomData);
+
+    DetectAppLayerMpmRegister2("tls.random", SIG_FLAG_TOCLIENT, 2, PrefilterGenericMpmRegister,
+            GetRandomData, ALPROTO_TLS, 0);
+
+    DetectBufferTypeSetDescriptionByName("tls.random", "TLS Random");
+
+    g_tls_random_buffer_id = DetectBufferTypeGetByName("tls.random");
+}
+
+/**
+ * \brief this function setup the tls.random sticky buffer keyword used in the rule
+ *
+ * \param de_ctx   Pointer to the Detection Engine Context
+ * \param s        Pointer to the Signature to which the current keyword belongs
+ * \param str      Should hold an empty string always
+ *
+ * \retval 0  On success
+ * \retval -1 On failure
+ */
+static int DetectTlsRandomTimeSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectBufferSetActiveList(s, g_tls_random_time_buffer_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_TLS) < 0)
+        return -1;
+
+    return 0;
+}
+
+static int DetectTlsRandomBytesSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectBufferSetActiveList(s, g_tls_random_bytes_buffer_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_TLS) < 0)
+        return -1;
+
+    return 0;
+}
+
+static int DetectTlsRandomSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectBufferSetActiveList(s, g_tls_random_buffer_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_TLS) < 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *GetRandomTimeData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *f, const uint8_t flow_flags, void *txv,
+        const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        uint8_t random_buf;
+        const SSLState *ssl_state = (SSLState *)f->alstate;
+
+        if (flow_flags & STREAM_TOSERVER) {
+            random_buf = ssl_state->server_connp.random[0];
+        } else {
+            random_buf = ssl_state->client_connp.random[0];
+        }
+
+        /* First four bytes of random represent current UTC in the
+         * Unix epoch format so for now, first byte should never be 0 */
```

Comment:
but what if they are? We can't match on it then? Think this needs a flag or some other way to see if the field is filled.

Reply:
the field is under full attacker control, so no assumptions about the values of the bytes are valid